docs/cudf/source/developer_guide/pylibcudf.md

```diff
@@ -0,0 +1,113 @@
+
+To satisfy the goals of pylibcudf, we impose the following set of design principles:
+- Every public function or method should be `cpdef`ed. This allows it to be used in both Cython and Python code. This incurs some slight overhead over `cdef` functions, but we assume that this is acceptable because 1) the vast majority of users will be using pure Python rather than Cython, and 2) the overhead of a `cpdef` function over a `cdef` function is on the order of a nanosecond, while CUDA kernel launch overhead is on the order of a microsecond, so these function overheads should be washed out by typical usage of pylibcudf.
+- Every variable used should be strongly typed and either be a primitive type (int, float, etc) or a cdef class. Any enums in C++ should be mirrored using `cpdef enum`, which will create both a C-style enum in Cython and a PEP 435-style Python enum that will automatically be used in Python.
+- All typing in code should be written using Cython syntax, not PEP 484 Python typing syntax. Not only does this ensure compatibility with Cython < 3, but even with Cython 3 PEP 484 support remains incomplete as of this writing.
+- All cudf code should interact only with pylibcudf, never with libcudf directly.
+- All imports should be relative so that pylibcudf can be easily extracted from cudf later
+  - Exception: All imports of libcudf API bindings in `cudf._lib.cpp` should use absolute imports of `cudf._lib.cpp as libcudf`. We should convert the `cpp` directory into a proper package so that it can be imported as `libcudf` in that fashion. When moving pylibcudf into a separate package, it will be renamed to `libcudf` and only the imports will need to change.
+- Ideally, pylibcudf should depend on nothing other than rmm and pyarrow. This will allow it to be extracted into a a largely standalone library and used in environments where the larger dependency tree of cudf may be cumbersome.
+
+
+## Relationship to libcudf
+
+In general, the relationship between pylibcudf and libcudf can be understood in terms of two components, data structures and algorithms.
+
+(data-structures)=
+
+### Data Structures
+
+Typically, every type in libcudf should have a mirror Cython `cdef` class with an attribute `self.c_obj: unique_ptr[${underlying_type}]` that owns an instance of the underlying libcudf type.
+Each type should also implement a corresponding method `cdef ${cython_type} from_${underlying_type}(${underlying_type} dt)` to enable constructing the Cython object from ann underlying libcudf instance.
+Depending on the nature of the type, the function may need to accept a `unique_ptr` and take ownership (TODO: This will typically be the case for types that own GPU data, may want to codify further).
+
+For example, `libcudf::data_type` maps to `pylibcudf.DataType`, which looks like this (implementation omitted):
+```cython
+
+cdef class DataType:
+    cdef data_type c_obj
+
+    cpdef TypeId id(self)
+    cpdef int32_t scale(self)
```

Comment:
Would it (in the interest of pythonicity) make sense to convert zero-argument method accessors into properties (or cached_properties)

Reply:
I considered that, but decided against it for two reasons:
1. Properties are pure Python, so they'll introduce nonzero overhead
2. Since properties are Python, there's no way to type them, so the resulting code ends up being equally verbose because the call sites now need to cdef new types to tell Cython the types of the outputs.

Reply:
Ah, thanks.